BFD Insights: Adding BCDA project

orenfromberg opened the pull request:
<!--

--- PR Hygiene Checklist ---

1. Make sure the changeset can be reviewed, keep it's scope and size succinct 
2. Make sure your branch is from your fork and has a meaningful name
3. Update the PR title: `BLUEBUTTON-99999 Add Awesomeness`
4. Edit the text below - do not leave placeholders in the text.
4.1. Remove sections that you don't feel apply
5. Add any other details that will be helpful for the reviewers: details description, screenshots, etc
6. Request a review from someone/multiple someones
7. <optional> Review your changes yourself and write up any comments / concerns as if you were reviewing someone else's code.
-->

### Change Details

<!-- Add detailed discussion of changes here: -->
<!-- This is likely a summary, or the complete contents, of your commit messages -->

This branch adds folder `terraform/projects/bcda` with some infrastructure that will be used to collect data for BFD insights.

The infrastructure consists of a terraform backend and an instance of the bucket module which creates a number of resources.

This module is copied almost exactly from the AB2D folder and allows for 1 cross account role to access the bucket.

Here is the terraform plan to be applied:
```
------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.bucket.aws_iam_group_policy_attachment.athena_attach[0] will be created
  + resource "aws_iam_group_policy_attachment" "athena_attach" {
      + group      = "bfd-insights-readers"
      + id         = (known after apply)
      + policy_arn = (known after apply)
    }

  # module.bucket.aws_iam_group_policy_attachment.athena_attach[1] will be created
  + resource "aws_iam_group_policy_attachment" "athena_attach" {
      + group      = "bfd-insights-authors"
      + id         = (known after apply)
      + policy_arn = (known after apply)
    }

  # module.bucket.aws_iam_group_policy_attachment.full_attach[0] will be created
  + resource "aws_iam_group_policy_attachment" "full_attach" {
      + group      = "bfd-insights-analysts"
      + id         = (known after apply)
      + policy_arn = (known after apply)
    }

  # module.bucket.aws_iam_policy.athena_query will be created
  + resource "aws_iam_policy" "athena_query" {
      + arn         = (known after apply)
      + description = "Rights needed for athena query access"
      + id          = (known after apply)
      + name        = "bfd-insights-athena-query-bcda"
      + path        = "/bfd-insights/"
      + policy      = (known after apply)
    }

  # module.bucket.aws_iam_policy.full will be created
  + resource "aws_iam_policy" "full" {
      + arn         = (known after apply)
      + description = "Allow full access and use of the bcda bucket"
      + id          = (known after apply)
      + name        = "bfd-insights-full-bcda"
      + path        = "/bfd-insights/"
      + policy      = (known after apply)
    }

  # module.bucket.aws_kms_alias.main will be created
  + resource "aws_kms_alias" "main" {
      + arn            = (known after apply)
      + id             = (known after apply)
      + name           = "alias/bfd-insights-bcda-cmk"
      + target_key_arn = (known after apply)
      + target_key_id  = (known after apply)
    }

  # module.bucket.aws_kms_grant.cmk_grant will be created
  + resource "aws_kms_grant" "cmk_grant" {
      + grant_id          = (known after apply)
      + grant_token       = (known after apply)
      + grantee_principal = "arn:aws:iam::577373831711:role/service-role/aws-quicksight-service-role-v0"
      + id                = (known after apply)
      + key_id            = (known after apply)
      + name              = "bfd-insights-bcda-cmk-grant"
      + operations        = [
          + "Decrypt",
          + "DescribeKey",
          + "Encrypt",
          + "GenerateDataKey",
          + "ReEncryptFrom",
          + "ReEncryptTo",
        ]
      + retire_on_delete  = false
    }

  # module.bucket.aws_kms_key.main will be created
  + resource "aws_kms_key" "main" {
      + arn                      = (known after apply)
      + customer_master_key_spec = "SYMMETRIC_DEFAULT"
      + description              = "CMK for the bfd-insights-bcda-577373831711 bucket"
      + enable_key_rotation      = true
      + id                       = (known after apply)
      + is_enabled               = true
      + key_id                   = (known after apply)
      + key_usage                = "ENCRYPT_DECRYPT"
      + policy                   = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "kms:*"
                      + Effect    = "Allow"
                      + Principal = {
                          + AWS = "arn:aws:iam::577373831711:root"
                        }
                      + Resource  = "*"
                      + Sid       = "AllowAdmin"
                    },
                  + {
                      + Action    = [
                          + "kms:ReEncrypt*",
                          + "kms:GenerateDataKey*",
                          + "kms:Encrypt",
                          + "kms:DescribeKey",
                          + "kms:Decrypt",
                        ]
                      + Effect    = "Allow"
                      + Principal = {
                          + AWS = "arn:aws:iam::755619740999:role/BcdaBfdInsightsRole"
                        }
                      + Resource  = "*"
                      + Sid       = "AllowUse"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + tags                     = {
          + "application" = "bfd-insights"
          + "business"    = "OEDA"
          + "project"     = "bcda"
        }
    }

  # module.bucket.aws_s3_bucket.main will be created
  + resource "aws_s3_bucket" "main" {
      + acceleration_status         = (known after apply)
      + acl                         = "private"
      + arn                         = (known after apply)
      + bucket                      = "bfd-insights-bcda-577373831711"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags                        = {
          + "application" = "bfd-insights"
          + "business"    = "OEDA"
          + "project"     = "bcda"
          + "sensitivity" = "moderate"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + server_side_encryption_configuration {
          + rule {
              + apply_server_side_encryption_by_default {
                  + kms_master_key_id = (known after apply)
                  + sse_algorithm     = "aws:kms"
                }
            }
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }
    }

  # module.bucket.aws_s3_bucket_object.top["databases"] will be created
  + resource "aws_s3_bucket_object" "top" {
      + acl                    = "private"
      + bucket                 = (known after apply)
      + content_type           = "application/x-directory"
      + etag                   = (known after apply)
      + force_destroy          = false
      + id                     = (known after apply)
      + key                    = "databases/"
      + server_side_encryption = (known after apply)
      + storage_class          = (known after apply)
      + version_id             = (known after apply)
    }

  # module.bucket.aws_s3_bucket_policy.cross_account[0] will be created
  + resource "aws_s3_bucket_policy" "cross_account" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

  # module.bucket.aws_s3_bucket_public_access_block.main will be created
  + resource "aws_s3_bucket_public_access_block" "main" {
      + block_public_acls       = true
      + block_public_policy     = true
      + bucket                  = (known after apply)
      + id                      = (known after apply)
      + ignore_public_acls      = true
      + restrict_public_buckets = true
    }

Plan: 12 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------
```

### Acceptance Validation

- [ ] Bucket is created and accessible by `arn:aws:iam::755619740999:role/BcdaBfdInsightsRole`

### Feedback Requested

All feedback is welcome.

### External References

- [BCDA-3733](https://jira.cms.gov/browse/BCDA-3733)

### Security Implications

- [ ] new software dependencies

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

orenfromberg commented:
relates to https://github.com/CMSgov/bcda-ops/pull/565

msnook commented on insights/terraform/projects/bcda/main.tf:
```diff
@@ -0,0 +1,29 @@
+data "aws_caller_identity" "current" {}
+
+
+## Bucket for the project's data
```
Will there be a single bucket for data coming from `prod` and a separate bucket for data coming from `opensbx`?  Or will all data flow into one bucket?

orenfromberg replied:
We can have a separate Kinesis Firehose Datastream for each environment but the data will all go in the same bucket. The buckets are only distinguished by sensitivity:
* **high**: Contains information which are highly sensitive. PII data has high sensitivity.
* **moderate**: Contains data which are moderately sensitive. DASG logs have moderate sensitivity. Logs are scrubbed not to contain high sensitivity data.

msnook reviewed (approved):
👍

orenfromberg commented:
Here is the terraform plan for adding the athena workgroup:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.workgroup.aws_athena_workgroup.main will be created
  + resource "aws_athena_workgroup" "main" {
      + arn           = (known after apply)
      + force_destroy = false
      + id            = (known after apply)
      + name          = "bcda"
      + state         = "ENABLED"
      + tags          = {
          + "application" = "bfd-insights"
          + "business"    = "OEDA"
          + "project"     = "bcda"
        }

      + configuration {
          + enforce_workgroup_configuration    = true
          + publish_cloudwatch_metrics_enabled = true

          + result_configuration {
              + output_location = "s3://bfd-insights-bcda-577373831711/workgroups/bcda/"

              + encryption_configuration {
                  + encryption_option = "SSE_KMS"
                  + kms_key_arn       = "arn:aws:kms:us-east-1:577373831711:key/a4ca59d6-3978-434b-9a5c-c6ae8896db18"
                }
            }
        }
    }

  # module.workgroup.aws_s3_bucket_object.folder will be created
  + resource "aws_s3_bucket_object" "folder" {
      + acl                    = "private"
      + bucket                 = "bfd-insights-bcda-577373831711"
      + content_type           = "application/x-directory"
      + etag                   = (known after apply)
      + force_destroy          = false
      + id                     = (known after apply)
      + key                    = "workgroups/bcda/"
      + server_side_encryption = (known after apply)
      + storage_class          = (known after apply)
      + version_id             = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: no

Apply cancelled.

oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket [!] via 💠 default on ☁️  bfd(us-east-1) took 29s 
[ 08:30:04 ] ❯ git add .                                                                                                                                                                            <aws:bfd>

oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket [+] via 💠 default on ☁️  bfd(us-east-1) 
[ 12:12:53 ] ❯ git status                                                                                                                                                                           <aws:bfd>
On branch oren/3733-insights-bucket
Your branch is up to date with 'origin/oren/3733-insights-bucket'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   main.tf


oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket [+] via 💠 default on ☁️  bfd(us-east-1) 
[ 12:12:54 ] ❯ git commit -m "Adding athena workgroup"                                                                                                                                              <aws:bfd>
[oren/3733-insights-bucket 1753a9e1] Adding athena workgroup
 1 file changed, 8 insertions(+)

oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket [⇡] via 💠 default on ☁️  bfd(us-east-1) took 3s 
[ 12:13:03 ] ❯ git push                                                                                                                                                                             <aws:bfd>
Host key fingerprint is SHA256:nThbg6kXUpJWGl7E1IGOCspRomTxdCARLviKw6E5SY8
+---[RSA 2048]----+
| =+o...+=o..     |
|o++... *o .      |
|*.o.  *o.        |
|oo.  ..o.= .     |
|.+o. .. S =      |
|*=+ .  o = .     |
|OE .  . o        |
| o     .         |
|                 |
+----[SHA256]-----+
Counting objects: 7, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 1.36 KiB | 1.36 MiB/s, done.
Total 7 (delta 4), reused 0 (delta 0)
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To github.com:CMSgov/beneficiary-fhir-data.git
   25294809..1753a9e1  oren/3733-insights-bucket -> oren/3733-insights-bucket

oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket via 💠 default on ☁️  bfd(us-east-1) took 3s 
[ 12:13:08 ] ❯ terraform apply                                                                                                                                                                      <aws:bfd>
Backend reinitialization required. Please run "terraform init".
Reason: Backend configuration changed for "s3"

The "backend" is the interface that Terraform uses to store state,
perform operations, etc. If this message is showing up, it means that the
Terraform configuration you're using is using a custom configuration for
the Terraform backend.

Changes to backend configurations require reinitialization. This allows
Terraform to setup the new configuration, copy existing state, etc. This is
only done during "terraform init". Please run that command now then try again.

If the change reason above is incorrect, please verify your configuration
hasn't changed and try again. At this point, no changes to your existing
configuration or state have been made.


Error: Initialization required. Please see the error message above.



oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket via 💠 default on ☁️  bfd(us-east-1) 
[ 12:13:12 ] ❯ tfswitch -b ~/.bin/terraform                                                                                                                                                         <aws:bfd>
✔ 0.13.3 *recent
Switched terraform to version "0.13.3" 

oren on 💻 oren-XPS-15-9560 in terraform/projects/bcda on  oren/3733-insights-bucket via 💠 default on ☁️  bfd(us-east-1) 
[ 12:13:20 ] ❯ terraform plan                                                                                                                                                                       <aws:bfd>
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_caller_identity.current: Refreshing state... [id=2020-09-30 12:29:11.84467125 +0000 UTC]
module.workgroup.data.aws_caller_identity.current: Refreshing state...
module.bucket.data.aws_caller_identity.current: Refreshing state... [id=2020-09-30 12:29:11.849477065 +0000 UTC]
data.aws_kms_alias.moderate_cmk: Refreshing state... [id=2020-09-30 12:29:11.898772719 +0000 UTC]
data.aws_s3_bucket.moderate_bucket: Refreshing state... [id=bfd-insights-moderate-577373831711]
module.bucket.data.aws_iam_policy_document.cmk_policy: Refreshing state... [id=164700154]
module.bucket.aws_kms_key.main: Refreshing state... [id=a4ca59d6-3978-434b-9a5c-c6ae8896db18]
module.bucket.aws_kms_alias.main: Refreshing state... [id=alias/bfd-insights-bcda-cmk]
module.workgroup.data.aws_kms_key.bucket_cmk: Refreshing state...
module.bucket.aws_kms_grant.cmk_grant: Refreshing state... [id=a4ca59d6-3978-434b-9a5c-c6ae8896db18:9fd7dbddd07d00c8915f8b71f92e47bc2c1e6b58a33b70effd3f977133be6c2e]
module.bucket.aws_s3_bucket.main: Refreshing state... [id=bfd-insights-bcda-577373831711]
module.bucket.aws_s3_bucket_public_access_block.main: Refreshing state... [id=bfd-insights-bcda-577373831711]
module.bucket.aws_s3_bucket_policy.cross_account[0]: Refreshing state... [id=bfd-insights-bcda-577373831711]
module.workgroup.data.aws_s3_bucket.bucket: Refreshing state...
module.bucket.aws_iam_policy.athena_query: Refreshing state... [id=arn:aws:iam::577373831711:policy/bfd-insights/bfd-insights-athena-query-bcda]
module.bucket.aws_s3_bucket_object.top["databases"]: Refreshing state... [id=databases/]
module.bucket.aws_iam_policy.full: Refreshing state... [id=arn:aws:iam::577373831711:policy/bfd-insights/bfd-insights-full-bcda]
module.bucket.aws_iam_group_policy_attachment.full_attach[0]: Refreshing state... [id=bfd-insights-analysts-20200924145631225200000002]
module.bucket.aws_iam_group_policy_attachment.athena_attach[1]: Refreshing state... [id=bfd-insights-authors-20200924145631229200000003]
module.bucket.aws_iam_group_policy_attachment.athena_attach[0]: Refreshing state... [id=bfd-insights-readers-20200924145631172200000001]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.workgroup.aws_athena_workgroup.main will be created
  + resource "aws_athena_workgroup" "main" {
      + arn           = (known after apply)
      + force_destroy = false
      + id            = (known after apply)
      + name          = "bcda"
      + state         = "ENABLED"
      + tags          = {
          + "application" = "bfd-insights"
          + "business"    = "OEDA"
          + "project"     = "bcda"
        }

      + configuration {
          + enforce_workgroup_configuration    = true
          + publish_cloudwatch_metrics_enabled = true

          + result_configuration {
              + output_location = "s3://bfd-insights-bcda-577373831711/workgroups/bcda/"

              + encryption_configuration {
                  + encryption_option = "SSE_KMS"
                  + kms_key_arn       = "arn:aws:kms:us-east-1:577373831711:key/a4ca59d6-3978-434b-9a5c-c6ae8896db18"
                }
            }
        }
    }

  # module.workgroup.aws_s3_bucket_object.folder will be created
  + resource "aws_s3_bucket_object" "folder" {
      + acl                    = "private"
      + bucket                 = "bfd-insights-bcda-577373831711"
      + content_type           = "application/x-directory"
      + etag                   = (known after apply)
      + force_destroy          = false
      + id                     = (known after apply)
      + key                    = "workgroups/bcda/"
      + server_side_encryption = (known after apply)
      + storage_class          = (known after apply)
      + version_id             = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.
```

orenfromberg commented:
Here is the terraform plan for adding the athena workgroup:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.workgroup.aws_athena_workgroup.main will be created
  + resource "aws_athena_workgroup" "main" {
      + arn           = (known after apply)
      + force_destroy = false
      + id            = (known after apply)
      + name          = "bcda"
      + state         = "ENABLED"
      + tags          = {
          + "application" = "bfd-insights"
          + "business"    = "OEDA"
          + "project"     = "bcda"
        }

      + configuration {
          + enforce_workgroup_configuration    = true
          + publish_cloudwatch_metrics_enabled = true

          + result_configuration {
              + output_location = "s3://bfd-insights-bcda-577373831711/workgroups/bcda/"

              + encryption_configuration {
                  + encryption_option = "SSE_KMS"
                  + kms_key_arn       = "arn:aws:kms:us-east-1:577373831711:key/a4ca59d6-3978-434b-9a5c-c6ae8896db18"
                }
            }
        }
    }

  # module.workgroup.aws_s3_bucket_object.folder will be created
  + resource "aws_s3_bucket_object" "folder" {
      + acl                    = "private"
      + bucket                 = "bfd-insights-bcda-577373831711"
      + content_type           = "application/x-directory"
      + etag                   = (known after apply)
      + force_destroy          = false
      + id                     = (known after apply)
      + key                    = "workgroups/bcda/"
      + server_side_encryption = (known after apply)
      + storage_class          = (known after apply)
      + version_id             = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.
```

keithdadkins reviewed (approved):
LGTM. I also see the tfstate file in bfd-insights/bcda.
